### PR TITLE
Adds AreMultiPartiesFullTeams and Test_BattlersShareParty handling to tests

### PR DESCRIPTION
**Pull request by grintoul1**
## Description
Adds AreMultiPartiesFullTeams and Test_BattlersShareParty handling to tests

## Discord contact info
grintoul

### CODE REVIEW

File: src/battle_ai_util.c [hunk 1]
```diff
@@ -6377,7 +6377,7 @@ bool32 AI_OpponentCanFaintAiWithMod(enum BattlerId battler, u32 healAmount)
 
 s32 GetAILastPartyIndex(enum BattlerId battler)
 {
-    return (BattleSideHasTwoTrainers(battler & BIT_SIDE) && !AreMultiPartiesFullTeams()) ? PARTY_SIZE / 2 : PARTY_SIZE;
+    return (gBattleStruct->battlerFullTeam >> battler) ? PARTY_SIZE : MULTI_PARTY_SIZE;
 }
```
Comment (AlexOn1ine) [inline src/battle_ai_util.c:6390]: I'm not sure If I agree overall with this PR. 

The value returned from `GetAILastPartyIndex` could be passed around as a variable to prevent recalculations and would probably be overall faster then calling the function each time. Afaik that's what @Pawkkie plans to do. However if that doesn't help we could look further into this

We still would end up with roughly the same saving if not more.

Generally speaking I'm not a fan of storing values into global variables that we can get with a function because the complexity* increases, unless 1. it's local or 2. it very critical.  

*can be seen in this PR. Instead of one single entry point there are now 2 additional functions and the code size almost tripled.

Comment (grintoul1) [reply src/battle_ai_util.c:6390]: Gotcha - as always open to any and all suggestions; seeking a way to not have to always do the two trainer + flags check for the last index

The BattleTrainer stuff is again trying to reduce the amount of stuff that has to be run because previously GetBattlerTrainer was quite messy and had different handling between live battles and the test system battles

File: src/battle_controllers.c [hunk 2]
```diff
@@ -3296,26 +3299,11 @@ void FreeShinyStars(void)
...
-        default:
-            break;
-        }
-    }
-
-    return (enum BattleTrainer)(BattleSideHasTwoTrainers(battler & BIT_SIDE) ? battler : battler & BIT_SIDE);
+#if TESTING   
+    if (!gMain.inBattle)         
+        return (gBattleTestRunnerState->data.battlerTrainers >> (battler * 2)) & 0x3;
+#endif
```
Comment (mrgriffin) [inline src/battle_controllers.c]: I thiiiink we might not need this case—not 100% sure, but [`Test_GetBattlerTrainer`](<https://github.com/rh-hideout/pokeemerald-expansion/blob/upcoming/test/test_runner_battle.c#L261>) exists and I'm not sure if there are other calls to `GetBattlerTrainer` from tests outside of the battle?

File: src/battle_util.c [hunk 3]
```diff
@@ -9725,7 +9726,29 @@ bool32 AreMultiPartiesFullTeams(void)
 {
     enum DifficultyLevel difficulty = GetCurrentDifficultyLevel();
 
-    if (B_MULTI_HALF_TEAMS
+    if (TESTING)
```
Comment (mrgriffin) [inline src/battle_util.c]: Just a thought, should this be a case for recorded battles rather than specifically tests? Or thinking about it another way, would this function return the right thing if you had a recorded (non?) link battle? Maybe the answer is "recorded battles are always full parties"? :)

File: src/battle_controllers.c (hunk 2 above)
Comment (grintoul1) [reply src/battle_controllers.c]: `BattlersShareParty` relies on comparing the result of `GetBattlerTrainer` meaning you get `SEND_OUT to battler` errors as Trainer(party) is seen as zero during the check

File: src/battle_util.c (hunk 3 above)
Comment (grintoul1) [reply src/battle_util.c]: >would this function return the right thing if you had a recorded (non?) link battle?

I want to say yes... I may be missing information, but I think the only other recorded battles are in the frontier? If so, the only time this check would be relevant is tower multi mode, which has a check in the `else if` branch below.

>Maybe the answer is "recorded battles are always full parties"

We can definitely limit it to `if (IsAITest())`

**Comment by grintoul1** [comment]
Drafting while options are discussed 🙂

File: src/battle_controllers.c (hunk 2 above)
Comment (mrgriffin) [reply src/battle_controllers.c]: I see! That sounds to me like we should have a test-specific check in the `SEND_OUT` code rather than calling `BattlesShareParty`. imo `SEND_OUT`'s checking isn't really a question about the state of the battle, it's a question about what the test runner thinks the state of the battle will be.

(It would be nice if it could be a question about the state of the battle, but that's future work because it would mean alternating between the test set-up and the battle engine.)

File: src/battle_util.c (hunk 3 above)
Comment (mrgriffin) [reply src/battle_util.c]: I think you're right that that's the only other source of recorded battles in vanilla. I guess downstream could try and record other types of battles, but I wouldn't worry about that—there's probably other things that are a bit off in that case too.

File: src/battle_controllers.c (hunk 2 above)
Comment (grintoul1) [reply src/battle_controllers.c]: Okay, I can replace `BattlersShareParty(battlerId, i)` with `(Test_GetBattlerTrainer(battlerId) == Test_GetBattlerTrainer(i))`? That's exactly what that check does, just that `BattlersShareParty` uses the non-test version

Comment (grintoul1) [reply src/battle_controllers.c]: in fact it's used 4 times so I'll make a `Test_BattlersShareParty`

File: src/battle_ai_util.c (hunk 1 above)
Comment (Pawkkie) [reply src/battle_ai_util.c:6390]: > The value returned from `GetAILastPartyIndex` could be passed around as a variable to prevent recalculations and would probably be overall faster then calling the function each time. Afaik that's what [@Pawkkie](https://github.com/Pawkkie) plans to do. However if that doesn't help we could look further into this

I am not qualified to talk about the affect on the test system, but I can confirm that storing that result in `struct SwitchContext` for reuse is part of the `ShouldSwitch` refactor that should be PR'd later this week. It'll cut down like 6 calls or so iirc, but I was just saving it locally to `ShouldSwitch`. Might make more sense to save it to the battle struct / thinking struct / logic data if it needs to be reused far outside that scope and is expensive.

**Review by AlexOn1ine — APPROVED** [review]

**Comment by AlexOn1ine** [comment]
@mrgriffin merge or dismiss please

**Comment by mrgriffin** [comment]
Sorry for the massive hold-up.

**Comment by grintoul1** [comment]
All good, thanks for the second set of eyes :)
